service/listeners.go

```diff
@@ -0,0 +1,290 @@
+	Close() error
+	Len() int
+}
+
+type listenerSet struct {
+	manager   ListenerManager
+	listeners map[string]Listener
+}
+
+func (ls *listenerSet) Listen(network string, addr string) (net.Listener, error) {
+	lnKey := listenerKey(network, addr)
+	if _, exists := ls.listeners[lnKey]; exists {
+		return nil, fmt.Errorf("listener %s already exists", lnKey)
+	}
+	ln, err := ls.manager.Listen(network, addr)
+	if err != nil {
+		return nil, err
+	}
+	ls.listeners[lnKey] = ln
+	return ln, nil
+}
+
+func (ls *listenerSet) ListenPacket(network string, addr string) (net.PacketConn, error) {
+	lnKey := listenerKey(network, addr)
+	if _, exists := ls.listeners[lnKey]; exists {
+		return nil, fmt.Errorf("listener %s already exists", lnKey)
+	}
+	ln, err := ls.manager.ListenPacket(network, addr)
+	if err != nil {
+		return nil, err
+	}
+	ls.listeners[lnKey] = ln
+	return ln, nil
+}
+
+func (ls *listenerSet) Close() error {
+	for _, listener := range ls.listeners {
+		switch ln := listener.(type) {
+		case net.Listener:
+			if err := ln.Close(); err != nil {
+				return fmt.Errorf("%s listener on address %s failed to stop: %w", ln.Addr().Network(), ln.Addr().String(), err)
+			}
+		case net.PacketConn:
+			if err := ln.Close(); err != nil {
+				return fmt.Errorf("%s listener on address %s failed to stop: %w", ln.LocalAddr().Network(), ln.LocalAddr().String(), err)
+			}
+		default:
+			return fmt.Errorf("unknown listener type: %v", ln)
+		}
+	}
+	return nil
+}
+
+func (ls *listenerSet) Len() int {
+	return len(ls.listeners)
+}
+
+// ListenerManager holds and manages the state of shared listeners.
+type ListenerManager interface {
+	NewListenerSet() ListenerSet
+	Listen(network string, addr string) (net.Listener, error)
+	ListenPacket(network string, addr string) (net.PacketConn, error)
+}
+
+type listenerManager struct {
+	listeners   map[string]*globalListener
+	listenersMu sync.Mutex
+}
+
+func NewListenerManager() ListenerManager {
+	return &listenerManager{
+		listeners: make(map[string]*globalListener),
+	}
+}
+
+func (m *listenerManager) NewListenerSet() ListenerSet {
+	return &listenerSet{
+		manager:   m,
+		listeners: make(map[string]Listener),
+	}
+}
+
+// ListenStream creates a new stream listener for a given network and address.
+//
+// Listeners can overlap one another, because during config changes the new
+// config is started before the old config is destroyed. This is done by using
+// reusable listener wrappers, which do not actually close the underlying socket
+// until all uses of the shared listener have been closed.
+func (m *listenerManager) Listen(network string, addr string) (net.Listener, error) {
+	m.listenersMu.Lock()
+	defer m.listenersMu.Unlock()
+
+	lnKey := listenerKey(network, addr)
+	if lnGlobal, ok := m.listeners[lnKey]; ok {
+		lnGlobal.usage.Add(1)
+		return &sharedListener{
+			listener: lnGlobal.ln,
+			usage:    &lnGlobal.usage,
+			acceptCh: lnGlobal.acceptCh,
+			closeCh:  make(chan struct{}),
+			closeFunc: func() {
```

Comment:
I'd call this `onCloseFunc`, since it's effectively an event handler, not really closing the listener.